### PR TITLE
progress bars for EnsembleProblem

### PR DESCRIPTION
**Pull request by pepijndevos**
This is the first part of a series of PRs that adds progress support to EnsemblePoblem.

What this part does is if progress is enabled, initialize progress bars for every trajectory, and give every solve an unique ID.

For very large ensembles the expectation is that the logger shows this in a sensible way, for example, show the overall progress:

```
sum_logger = let progress = Dict{Symbol, Float64}()
    TransformerLogger(TerminalLogger()) do log
        if log.level == LogLevel(-1) && haskey(log.kwargs, :progress)
            @show log
            pr = log.kwargs[:progress]
            if pr isa Number
                progress[log.id] = pr
            elseif pr == "done"
                progress[log.id] = 1.0
            end
            tot = sum(values(progress))/length(progress)
            if tot>=1.0
                tot="done"
                empty!(progress)
            end
            log = merge(log, (;id=:total, message="Total", kwargs=Dict(:progress=>tot)))
        else
            log
        end
    end
end
global_logger(sum_logger)
```

### CODE REVIEW

**Comment by devmotion** [comment]
> For very large ensembles the expectation is that the logger shows this in a sensible way, for example, show the overall progress:

Last time I checked none of the common loggers (the default logger in VSCode and TerminalLoggers) did summarize large numbers of progress bars. Everything would be displayed which lead to a terrible user experience IMO.

**Comment by pepijndevos** [comment]
Yea that's a fair criticism. I did it like this for a few reasons:

- The alternative would be some complicated thing with channels
- For small ensembles the current default is reasonable, and gives more insight than a single progress bar
- It's fairly easy to aggregate the logs if desired, as the snippet above demonstrates

Maybe it'd be a good idea to contribute this log aggregation somewhere.
We could even have that as the default where in `__solve` we just do `with_logger(sum_logger(current_logger()))`

**Comment by devmotion** [comment]
I'd like to have better progress bars for ensemble simulations (not only for SciML but e.g. also in Turing where we currently only use a single progress bar for multi-chain sampling which is updated only when a full chain is sampled and hence doesn't either lead to a good user experience) but I think it's crucial to first add better support for such summaries, ideally not only in SciML but more generally for the most common loggers such as the one in VSCode or TerminalLoggers (see e.g. https://github.com/julia-vscode/julia-vscode/issues/3297, and possibly also something like https://github.com/julia-vscode/julia-vscode/issues/3317 and https://github.com/JuliaLogging/ProgressLogging.jl/issues/23).

**Comment by oscardssmith** [comment]
A possible alternative approach here would be to make the ensembleprob's progress bar be purely based on the number of solutions in the ensemble that have finished.

**Comment by devmotion** [comment]
Yeah that's what we do in Turing currently. But the user experience is a bit suboptimal, in particular if you only sample a few chains in parallel - then it can happen that the progress bar is at 0 all the time and jumps to 1 when all chains are sampled at approx the same time.

File: src/ensemble/basic_ensemble_solve.jl [hunk 1]
```diff
@@ -97,7 +105,12 @@ function batch_func(i, prob, alg; kwargs...)
     _prob = prob.safetycopy ? deepcopy(prob.prob) : prob.prob
     new_prob = prob.prob_func(_prob, i, iter)
     rerun = true
-    x = prob.output_func(solve(new_prob, alg; kwargs...), i)
+
+    name = get(kwargs, :progress_name, "ODE")
```
Comment (ChrisRackauckas) [inline src/ensemble/basic_ensemble_solve.jl:109]: This isn't necessarily an ODE

File: src/ensemble/basic_ensemble_solve.jl [hunk 2]
```diff
@@ -60,6 +61,13 @@ function __solve(prob::AbstractEnsembleProblem,
         error("trajectories ÷ batch_size cannot be less than 1, got $num_batches")
     num_batches * batch_size != trajectories && (num_batches += 1)
 
+    if get(kwargs, :progress, false)
+        name = get(kwargs, :progress_name, "ODE")
```
Comment (ChrisRackauckas) [inline src/ensemble/basic_ensemble_solve.jl:65]: not an ODE

**Comment by pepijndevos** [comment]
I proposed for Cedar that we'd do the progress of completed simulations but indeed the granularity is quit low if you have a lot of cores or not that many problems. So IMO the correct solution is for the loggers to handle this better, or to aggregate the results internally like I suggested.

File: src/ensemble/basic_ensemble_solve.jl [hunk 3]
```diff
@@ -109,7 +122,7 @@ function batch_func(i, prob, alg; kwargs...)
         iter += 1
         _prob = prob.safetycopy ? deepcopy(prob.prob) : prob.prob
         new_prob = prob.prob_func(_prob, i, iter)
-        x = prob.output_func(solve(new_prob, alg; kwargs...), i)
+        x = prob.output_func(solve(new_prob, alg; progress_name, progress_id, kwargs...), i)
```
Comment (ChrisRackauckas) [inline src/ensemble/basic_ensemble_solve.jl:125]: Are you able to setup a downstream weakdep so that this is forced to be on a version where it will work?

Comment (pepijndevos) [reply src/ensemble/basic_ensemble_solve.jl:125]: I don't think I can set a version constraint that works until there is a version with https://github.com/SciML/DiffEqBase.jl/pull/939 so that should be merged and released first so I can then constrain this to require that version.

**Comment by pepijndevos** [comment]
I've added an internal aggregation step that is enabled by default. This makes sure the current UX isn't horrible, and then we can take it from there.

**Comment by pepijndevos** [comment]
It turns out that if you have many cores and you set `progress_steps` too low it can cause quite a lot of lock contention, so I made the aggregator use `trylock` to mitigate that.

File: src/ensemble/basic_ensemble_solve.jl [hunk 4]
```diff
@@ -53,51 +95,66 @@ end
...
-        error("trajectories ÷ batch_size cannot be less than 1, got $num_batches")
-    num_batches * batch_size != trajectories && (num_batches += 1)
-
-    if num_batches == 1 && prob.reduction === DEFAULT_REDUCTION
-        elapsed_time = @elapsed u = solve_batch(prob, alg, ensemblealg, 1:trajectories,
-            pmap_batch_size; kwargs...)
-        _u = tighten_container_eltype(u)
-        return EnsembleSolution(_u, elapsed_time, true)
-    end
+    logger = progress_aggregate ? AggregateLogger(current_logger()) : current_logger()
```
Comment (ChrisRackauckas) [inline src/ensemble/basic_ensemble_solve.jl:100]: does this cause dynamic dispatch?

**Comment by pepijndevos** [comment]
FTR we're still investigating how to reduce overhead from progress logging, so some changes are still expected.

File: src/ensemble/basic_ensemble_solve.jl (hunk 4 above)
Comment (pepijndevos) [reply src/ensemble/basic_ensemble_solve.jl:100]: Not any more than `current_logger` already does
```
julia> Base.return_types(current_logger, Tuple{})
1-element Vector{Any}:
 Base.CoreLogging.AbstractLogger

julia> test(cond) = cond ? ConsoleLogger() : current_logger()
test (generic function with 1 method)

julia> Base.return_types(test, Tuple{Bool})
1-element Vector{Any}:
 Base.CoreLogging.AbstractLogger
```

File: src/ensemble/basic_ensemble_solve.jl [hunk 5]
```diff
@@ -23,6 +24,47 @@ $(TYPEDEF)
...
+        else
+            return
+        end
+    end
+    Logging.handle_message(l.logger, level, message, _module, group, id, file, line; kwargs...)
+end
+Logging.shouldlog(l::AggregateLogger, args...) = Logging.shouldlog(l.logger, args...)
+Logging.min_enabled_level(l::AggregateLogger) = Logging.min_enabled_level(l.logger)
+Logging.catch_exceptions(l::AggregateLogger) = Logging.catch_exceptions(l.logger)
+
```
Comment (pepijndevos) [inline src/ensemble/basic_ensemble_solve.jl:67]: I tried this, but it seems to have weird scheduling problems where the progress bar doesn't complete until much later.

```suggestion
mutable struct AggregateLogger{T<:AbstractLogger} <: AbstractLogger
    ch::Channel{Tuple{Symbol, Union{String, Float64}}}
    logger::T
end
function AggregateLogger(logger::AbstractLogger)
    progress = Dict{Symbol, Float64}()
    last_print = 0.0
    num_tasks_complete = 0
    ch = Channel{Tuple{Symbol, Union{String, Float64}}}(Threads.nthreads(); spawn=true) do ch
        for (id, pr) in ch
            println(id, " ", pr)
            if pr == "done"
                num_tasks_complete += 1
                pr = 1.0
            end
            progress[id] = pr
            curr_time = time()
            if num_tasks_complete >= length(progress)
                tot="done"
                empty!(progress)
                num_tasks_complete = 0
                last_print = 0.0
            elseif curr_time-last_print > 0.1
                last_print = curr_time
                tot = mean(values(progress))
            else
                continue
            end
            Logging.handle_message(logger, LogLevel(-1), "Total", @__MODULE__, :progress, :total, @__FILE__, @__LINE__; progress=tot)
        end
    end
    AggregateLogger(ch, logger)
end

function Logging.handle_message(l::AggregateLogger, level, message, _module, group, id, file, line; kwargs...)
    if convert(LogLevel, level) == LogLevel(-1) && haskey(kwargs, :progress)
        pr = kwargs[:progress]
        if pr isa Number
            pr = Float64(pr)
        end
        put!(l.ch, (id, pr))
    else
        Logging.handle_message(l.logger, level, message, _module, group, id, file, line; kwargs...)
    end
end
Logging.shouldlog(l::AggregateLogger, args...) = Logging.shouldlog(l.logger, args...)
Logging.min_enabled_level(l::AggregateLogger) = Logging.min_enabled_level(l.logger)
Logging.catch_exceptions(l::AggregateLogger) = Logging.catch_exceptions(l.logger)
```

File: src/ensemble/basic_ensemble_solve.jl (hunk 4 above)
Comment (ChrisRackauckas) [reply src/ensemble/basic_ensemble_solve.jl:100]: Yes, which is why we don't use it.

Comment (pepijndevos) [reply src/ensemble/basic_ensemble_solve.jl:100]: What I mean is, any logging call is always going to be dynamic dispatch because the logger is context dependent and can be changed at runtime

**Comment by pepijndevos** [comment]
Alright, performance overhead is now negligible using several strategies:
- use `trylock` to avoid contention on non-critical updates
- use `time` to limit updates to ten per second to limit downstream logging overhead
- use math to avoid summing 20k items
- teach Sundial to respect `progress_steps`

Profile:
image

I've also added the requested weakdep on https://github.com/SciML/DiffEqBase.jl/releases/tag/v6.132.0

**Comment by ChrisRackauckas** [comment]
As long as https://github.com/SciML/DiffEqBase.jl/blob/master/test/downstream/inference.jl passes this should generally be good.

**Comment by pepijndevos** [comment]
I think that particular test seems to pass. Which of the other failures out of 70 test runs are my fault is hard for me to tell.

**Comment by ChrisRackauckas** [comment]
I rebased. We'll see how tests go.

File: Project.toml [hunk 6]
```diff
@@ -70,6 +71,7 @@ Tables = "1"
 TruncatedStacktraces = "1"
 ZygoteRules = "0.2"
 julia = "1.6"
+DiffEqBase = "6.132.0"
```
Comment (ChrisRackauckas) [inline Project.toml:74]: why is this required?

Comment (pepijndevos) [reply Project.toml:74]: Eh I thought you asked for this so that the `progress_id` is guaranteed to work.

https://github.com/SciML/SciMLBase.jl/pull/514#discussion_r1341953237

Comment (pepijndevos) [reply Project.toml:74]: Is that not what you wanted?

**Comment by ChrisRackauckas** [comment]
It looks like a backwards compat dispatch is missing: https://github.com/SciML/SciMLBase.jl/actions/runs/6479781336/job/17731842462?pr=512

File: Project.toml [hunk 7]
```diff
@@ -37,7 +37,6 @@ PyCall = "438e738f-606a-5dbb-bf0a-cddfbfd45ab0"
 PythonCall = "6099a3de-0909-46bc-b1f4-468b9a2dfc0d"
 RCall = "6f49c342-dc21-5d91-9882-a32aef131414"
 Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"
-DiffEqBase = "2b5f629d-d688-5b77-993f-72d75c75574e"
```
Comment (pepijndevos) [inline Project.toml]: I think removing this will also require removing the compat entry below, no? Basically revert 8518c79e4622f56e2270de50096c9bfe5f9438ef

**Comment by staticfloat** [comment]
@ChrisRackauckas What else needs to happen for this to get in?

**Comment by ChrisRackauckas** [comment]
The try/catch handling here is incompatible with Zygote https://github.com/SciML/SciMLBase.jl/actions/runs/6532475287/job/17735718809?pr=514#step:6:1447

**Comment by pepijndevos** [comment]
So what can we do? You need a lock for thread safe coordination and you need finally to release the lock. Do we need to register handle_message as a zygote primitive or something?

**Comment by ChrisRackauckas** [comment]
Even easier, just set ignore derivatives needs to be set on the logging calls so it ignores everything in the function.

Have you done regression testing to check the performance?

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/SciMLBase.jl/actions/runs/6532475287/job/17735719200?pr=514 is pointing out that the extra kwargs are causing issues. In order to for this to get in, every downstream solver package needs to support the new keyword arguments. This includes:

- StochasticDiffEq
- JumpProcesses
- DelayDiffEq
- ODEInterfaceDiffEq
- NonlinearSolve
- SimpleNonlinearSolve
- Optimization
- Integrals

It would be fine to set it so that if the value is not the default that you just warn that progress isn't implemented (behind a `if verbose)`) and make sure the kwarg is allowed.

**Comment by pepijndevos** [comment]
I've never worked with zygote, are there some docs about how to do the derivative or is it a one line thing that would take you less time than finding the docs?

We've tested performance in Cedar and the overhead from logging is minor. I go to quite some length to avoid blocking on a lock or sending more updates than necessary. Julia logging is also such that the message isn't even computed if logging is disabled, so there should be no impact unless you enable progress logging.

**Comment by ChrisRackauckas** [comment]
Can you show it on something simple like the Lorenz equation with SimpleTsit5?

**Comment by pepijndevos** [comment]
I'm looking at the docs and all I can find is how to define a rrule? https://fluxml.ai/Zygote.jl/latest/limitations/#Solutions-1
Feels like this should go in ChainRules then?
Or do you mean something else by
>  just set ignore derivatives needs to be set on the logging calls so it ignores everything in the function.

**Comment by pepijndevos** [comment]
Well that didn't work. What gives? https://github.com/SciML/SciMLBase.jl/actions/runs/6690536216/job/18176044626?pr=514#step:6:1470

**Comment by pepijndevos** [comment]
Keeping track of all the PRs:


- [ ]     StochasticDiffEq https://github.com/SciML/StochasticDiffEq.jl/pull/552
- [ ]     JumpProcesses
- [ ]     DelayDiffEq
- [ ]     ODEInterfaceDiffEq
- [ ]     NonlinearSolve
- [ ]     SimpleNonlinearSolve
- [ ]     Optimization
- [ ]     Integrals

**Comment by pepijndevos** [comment]
FTR I'm getting weird precompile errors about the adjoint so maybe that's still not quite right?
```
┌ DiffEqBase → DiffEqBaseZygoteExt
│  WARNING: Method definition adjoint(ZygoteRules.AContext, typeof(ZygoteRules.literal_getproperty), SciMLBase.EnsembleSolution{T, N, S} where S where N where T, Base.Val{:u}) in module DiffEqBase overwritten in module SciMLBaseZygoteExt.
│  ┌ Error: Error during loading of extension SciMLBaseZygoteExt of SciMLBase, use `Base.retry_load_extensions()` to retry.
│  │   exception =
│  │    1-element ExceptionStack:
│  │    Method overwriting is not permitted during Module precompile.
```

File: src/ensemble/basic_ensemble_solve.jl [hunk 8]
```diff
@@ -1,3 +1,5 @@
+using Logging, ChainRulesCore
```
Comment (oxinabox) [inline src/ensemble/basic_ensemble_solve.jl:1]: Possibly we need to delete these. idk SciML practices but I don't think the ChainRulesCore one at least is needed

Comment (ChrisRackauckas) [reply src/ensemble/basic_ensemble_solve.jl:1]: Definitely should not be in the file and needs to be at the top level. Also the ChainRulesCore one is not needed. And I think we already import logging?

**Comment by pepijndevos** [comment]
Am I going crazy?
https://github.com/SciML/SciMLBase.jl/actions/runs/6769050135/job/18394689941#step:6:579
https://github.com/pepijndevos/SciMLBase.jl/blob/pv/progress/src/SciMLBase.jl#L25

**Comment by pepijndevos** [comment]
I went back to the alternative logging workaround and did some more import fixes that were hidden by my repl state I guess. This passes the downstream AD tests for me, but I think it requires a new ChainRules release before it'll work on CI.

**Comment by oxinabox** [comment]
ok new chainrules release is out, so restarting CI and it should pass

**Comment by oscardssmith** [comment]
this pr needs to bump the chainrules version requirement

**Comment by pepijndevos** [comment]
Last time Chris asked for a version requirement I added it and then he removed it again so idk

**Comment by ChrisRackauckas** [comment]
I don't think this PR ever had a version requirement on ChainRules/ChainRulesCore?

**Comment by pepijndevos** [comment]
No I added one for DiffEqBase because I thought you wanted that to guarantee `progress_id` is supported but then you removed it again so I'm fine adding a version constraint if it's clear that that's actually what you want

**Comment by ChrisRackauckas** [comment]
I don't see why that's related. @oscardssmith is saying you need to version bound on ChainRules/Core since you're implicitly relying on @oxinabox 's latest release for the Zygote fix on logging.

**Comment by pepijndevos** [comment]
Done. Just wanted to make sure.

I'm still getting that weird DiffEqBaseZygoteExt precompile error btw

**Comment by pepijndevos** [comment]
Here are your performance numbers btw
```julia
using OrdinaryDiffEq, BenchmarkTools
function lorenz(u, p, t)
    σ = p[1]
    ρ = p[2]
    β = p[3]
    du1 = σ * (u[2] - u[1])
    du2 = u[1] * (ρ - u[3]) - u[2]
    du3 = u[1] * u[2] - β * u[3]
    return [du1, du2, du3]
end

u0 = [1.0f0; 0.0f0; 0.0f0]
tspan = (0.0f0, 10.0f0)
p = [10.0f0, 28.0f0, 8 / 3.0f0]
prob = ODEProblem{false}(lorenz, u0, tspan, p)

prob_func = (prob, i, repeat) -> remake(prob, p = rand(Float32, 3) .* p)
monteprob = EnsembleProblem(prob, prob_func = prob_func, safetycopy = false)
@benchmark sol = solve(monteprob, Tsit5(), EnsembleThreads(), trajectories = 10_000, saveat = 1.0f0, progress=false)
@benchmark sol = solve(monteprob, Tsit5(), EnsembleThreads(), trajectories = 10_000, saveat = 1.0f0, progress=true)
```
```
BenchmarkTools.Trial: 12 samples with 1 evaluation.
 Range (min … max):  186.853 ms … 720.231 ms  ┊ GC (min … max):  0.00% … 73.95%
 Time  (median):     500.877 ms               ┊ GC (median):    62.15%
 Time  (mean ± σ):   446.853 ms ± 206.921 ms  ┊ GC (mean ± σ):  57.52% ± 33.01%

  ██                       ▁         █ ▁    ▁             ▁ ▁ ▁  
  ██▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁█▁█▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁█▁█▁█ ▁
  187 ms           Histogram: frequency by time          720 ms <

 Memory estimate: 3.40 GiB, allocs estimate: 45697823.

BenchmarkTools.Trial: 10 samples with 1 evaluation.
 Range (min … max):  265.698 ms … 743.625 ms  ┊ GC (min … max):  0.00% … 64.56%
 Time  (median):     558.434 ms               ┊ GC (median):    52.23%
 Time  (mean ± σ):   516.208 ms ± 154.807 ms  ┊ GC (mean ± σ):  49.51% ± 23.63%

  █                    ▁         ▁    ▁▁▁   ▁      ▁          ▁  
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁█▁▁▁▁███▁▁▁█▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁█ ▁
  266 ms           Histogram: frequency by time          744 ms <

 Memory estimate: 3.55 GiB, allocs estimate: 47195479.
```
Logging output of 10k simulation runs after all these optimizations is just a hand full of lines
```
┌ LogLevel(-1): Total
│   progress = 0.0
└ @ SciMLBase ~/code/SciMLBase.jl/src/ensemble/basic_ensemble_solve.jl:134
┌ LogLevel(-1): Total
│   message = "dt=0.07805729\nt=10.0\nmax u=25.905659"
│   progress = 0.211799999999993
└ @ OrdinaryDiffEq ~/code/OrdinaryDiffEq.jl/src/integrators/integrator_utils.jl:153
┌ LogLevel(-1): Total
│   progress = 0.756299999999933
└ @ OrdinaryDiffEq ~/code/OrdinaryDiffEq.jl/src/solve.jl:103
┌ LogLevel(-1): Total
│   message = "dt=0.046221733\nt=10.0\nmax u=23.776852"
│   progress = "done"
└ @ OrdinaryDiffEq ~/code/OrdinaryDiffEq.jl/src/integrators/integrator_utils.jl:153
```

**Comment by pepijndevos** [comment]
For reference, on master
```
BenchmarkTools.Trial: 12 samples with 1 evaluation.
 Range (min … max):  189.615 ms … 701.238 ms  ┊ GC (min … max):  0.00% … 72.88%
 Time  (median):     493.090 ms               ┊ GC (median):    61.75%
 Time  (mean ± σ):   457.536 ms ± 160.567 ms  ┊ GC (mean ± σ):  58.25% ± 25.18%

  ██           █        █           ███  ██  █          █     █  
  ██▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁███▁▁██▁▁█▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁█ ▁
  190 ms           Histogram: frequency by time          701 ms <

 Memory estimate: 3.40 GiB, allocs estimate: 45635653.

BenchmarkTools.Trial: 1 sample with 1 evaluation.
 Single result which took 6.611 s (3.17% GC) to evaluate,
 with a memory estimate of 3.77 GiB, over 51008757 allocations.
```

thousands of log messages omitted for brevity ;)

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/SciMLBase.jl/actions/runs/6795944602/job/18475031150?pr=514 looks like the last issue.

**Comment by pepijndevos** [comment]
Ahhh I thought that must have been some CI glitch from the stats PR, but turns out that in one of the many rebases conflicts of this PR over its long lifetime the stats got lost.

**Comment by pepijndevos** [comment]
Seems fine now? Could it be. Could it really be done?

**Comment by ChrisRackauckas** [comment]
# Before

#=
BenchmarkTools.Trial: 198 samples with 1 evaluation.
 Range (min … max):  22.291 ms … 43.583 ms  ┊ GC (min … max): 0.00% … 43.89%
 Time  (median):     23.324 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   25.246 ms ±  5.548 ms  ┊ GC (mean ± σ):  6.91% ± 12.90%

  ▆█▇▇▅▃▂                                                      
  ███████▆▁▆▅▅▅▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅▁▁▅▁▁▁▁▁▁▁▁▅▁▁▅▆▆▆▆▁▇█ ▅
  22.3 ms      Histogram: log(frequency) by time      42.7 ms <

 Memory estimate: 24.42 MiB, allocs estimate: 320057.

BenchmarkTools.Trial: 89 samples with 1 evaluation.
 Range (min … max):  41.775 ms … 83.893 ms  ┊ GC (min … max):  0.00% … 34.04%
 Time  (median):     44.861 ms              ┊ GC (median):     0.00%
 Time  (mean ± σ):   56.267 ms ± 15.007 ms  ┊ GC (mean ± σ):  22.08% ± 19.77%

   █▄                                                          
  ████▇▄▃▁▁▁▁▁▁▁▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃▄▅▃▄▅▆▇▆▄▄▅▁▁▁▁▃▃ ▁
  41.8 ms         Histogram: frequency by time        80.1 ms <

 Memory estimate: 152.59 MiB, allocs estimate: 1190057.
=#

# After

#=
BenchmarkTools.Trial: 221 samples with 1 evaluation.
 Range (min … max):  19.745 ms … 51.209 ms  ┊ GC (min … max): 0.00% … 56.97%
 Time  (median):     20.502 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   22.704 ms ±  7.274 ms  ┊ GC (mean ± σ):  8.31% ± 14.29%

  ▇█▅▂▁                                                        
  █████▄▅▁▁▁▄▁▄▁▁▁▁▁▄▁▁▁▁▁▁▁▁▁▁▁▁▁▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅▄▇▄▅▁▆ ▅
  19.7 ms      Histogram: log(frequency) by time      50.6 ms <

 Memory estimate: 24.42 MiB, allocs estimate: 320076.

BenchmarkTools.Trial: 31 samples with 1 evaluation.
 Range (min … max):  141.556 ms … 179.966 ms  ┊ GC (min … max):  4.08% … 11.54%
 Time  (median):     166.417 ms               ┊ GC (median):    11.92%
 Time  (mean ± σ):   166.345 ms ±   8.579 ms  ┊ GC (mean ± σ):  12.01% ±  2.93%

                           ▃         █  ▃  ▃                  ▃  
  ▇▁▁▁▁▁▁▁▁▇▁▁▁▁▁▁▁▁▁▁▁▁▁▁▇█▁▁▁▁▇▇▁▇▇█▁▇█▇▇█▁▇▇▇▇▇▁▁▁▇▇▁▇▁▇▇▁▁█ ▁
  142 ms           Histogram: frequency by time          180 ms <

 Memory estimate: 220.28 MiB, allocs estimate: 1785509.
=#
